src/renderer/dx/DxFontRenderData.cpp

```diff
@@ -441,6 +453,201 @@ try
 }
 CATCH_RETURN()
 
+// Routine Description:
+// - Returns whether the user set or updated any of the font features to be applied
+bool DxFontRenderData::DidUserSetFeatures() const noexcept
+{
+    return _didUserSetFeatures;
+}
+
+// Routine Description:
+// - Updates our internal map of font features with the given features
+// - NOTE TO CALLER: Make sure to call UpdateFont after calling this for the feature changes
+//   to take place
+// Arguments:
+// - features - the features to update our map with
+void DxFontRenderData::SetFeatures(const std::unordered_map<std::wstring_view, uint32_t>& features)
+{
+    // Populate the feature map with the standard list first
+    std::unordered_map<DWRITE_FONT_FEATURE_TAG, uint32_t> featureMap{
+        { DWRITE_MAKE_FONT_FEATURE_TAG('r', 'l', 'i', 'g'), 1 }, // Required Ligatures
+        { DWRITE_MAKE_FONT_FEATURE_TAG('r', 'c', 'l', 't'), 1 }, // Required Contextual Alternates
+        { DWRITE_MAKE_FONT_FEATURE_TAG('l', 'o', 'c', 'l'), 1 }, // Localized Forms
+        { DWRITE_MAKE_FONT_FEATURE_TAG('c', 'c', 'm', 'p'), 1 }, // Glyph Composition / Decomposition
+        { DWRITE_MAKE_FONT_FEATURE_TAG('c', 'a', 'l', 't'), 1 }, // Contextual Alternates
+        { DWRITE_MAKE_FONT_FEATURE_TAG('l', 'i', 'g', 'a'), 1 }, // Standard Ligatures
+        { DWRITE_MAKE_FONT_FEATURE_TAG('c', 'l', 'i', 'g'), 1 }, // Contextual Ligatures
+        { DWRITE_MAKE_FONT_FEATURE_TAG('k', 'e', 'r', 'n'), 1 }, // Kerning
+        { DWRITE_MAKE_FONT_FEATURE_TAG('m', 'a', 'r', 'k'), 1 }, // Mark Positioning
+        { DWRITE_MAKE_FONT_FEATURE_TAG('m', 'k', 'm', 'k'), 1 }, // Mark to Mark Positioning
+        { DWRITE_MAKE_FONT_FEATURE_TAG('d', 'i', 's', 't'), 1 }  // Distances
+    };
+
+    // Update our feature map with the provided features

```

Comment:
i think we need to set `_didUserSetFeatures` to false here so that if the user sets _and then deletes_ features... it knows that they didn't add any

Reply:
Yes, thanks for the catch!